File: poo/interfaces/class.go
package main

import "fmt"

type Person struct {
	name string
	age  int
}

type Employee struct {
	id int
}

type FullTimeEmployee struct {
	Person
	Employee
	endDate string
}

func (ftEmployee FullTimeEmployee) getMessage() string {
	return "Full time employee"
}

type TemporaryEmployee struct {
	Person
	Employee
	taxRate int
}

func (tEmployee TemporaryEmployee) getMessage() string {
	return "Temporary employee"
}

type PrintInfo interface {
	getMessage() string
}

func getMessage(p PrintInfo) {
	fmt.Println(p.getMessage())
}

func main() {
	ftEmployee := FullTimeEmployee{}
	ftEmployee.name = "Rodo"
	ftEmployee.age = 32
	ftEmployee.id = 1
	fmt.Printf("%v\n", ftEmployee)
	tEmployee := TemporaryEmployee{}
	getMessage(tEmployee)
	getMessage(ftEmployee)
	//GetMessage(ftEmployee)
}

File: poo/inheritance/class.ts
class BaseEmployee {
  private name: string;
  private id: number;

  constructor(id: number, name: string) {
    this.id = id;
    this.name = name;
  }

  public getMessage(): string {
    return `${this.name} with Id ${this.id}`;
  }
  
  public getName(): string {
    return this.name;
  }

  public setName(name: string): void {
    this.name = name
  }

  public getId(): number {
    return this.id;
  }

  public setId(id: number): void {
    this.id = id;
  }
}

class TemporaryEmployee extends BaseEmployee {
  private endDate: string;

  constructor (id: number, name: string, endDate: string) {
    super(id,name);
    this.endDate = endDate;
  }

  public getEndDate(): string{
    return this.endDate;
  }

  public getMessage(): string {
    return `${this.getName()} with id: ${this.getId()} and endDate ${this.getEndDate()}`;
  }
}

class FullTimeEmployee extends BaseEmployee {
  private taxRate: number;

  constructor(id:number, name: string, taxRate: number) {
    super(id, name);
    this.taxRate = taxRate;
  }

  public getTaxRate(): number {
    return this.taxRate;
  }

  public getMessage(): string {
    return `${this.getName()} with id: ${this.getId()} and taxRate ${this.getTaxRate()}`;
  }
}

const printGetMessage = (employee: BaseEmployee): void => {
  console.log(employee.getMessage());
}

const ftEmployee = new FullTimeEmployee(1, 'Name', 5);
const tEmployee = new TemporaryEmployee(1, 'Name 2', 'July');

printGetMessage(ftEmployee);
printGetMessage(tEmployee);